File: 프로그래머스/2/87946. 피로도/피로도.py
answer = 0
def solution(k, dungeons):
    visited = []
    
    def dfs(k, clear): #피로도, 클리어
        global answer
        answer = max(answer, clear)
        
        for idx, dungeon in enumerate(dungeons):
            if idx in visited:
                continue
            if k < dungeon[0]:
                continue
                
            visited.append(idx)
            dfs(k-dungeon[1], clear+1)
            visited.pop()
    dfs(k, 0)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>

using namespace std;

int maxDepth = 0;
bool visited[8] = {false,};

void dfs(int depth, int k, vector<vector<int>> &dungeons) {
    if (maxDepth < depth)
        maxDepth = depth;
    
    for (int i=0; i<dungeons.size(); i++) {
        if (visited[i] or dungeons[i][0] > k)
            continue;
        
        visited[i] = true;
        dfs(depth+1, k-dungeons[i][1], dungeons);
        visited[i] = false;
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    dfs(0, k, dungeons);
    return maxDepth;
}